backend/services/inventory/api/http/routing.go

```diff
@@ -38,27 +46,124 @@ func supportsMethod(method string, methods []string) bool {
 
 // Automatically add OPTIONS method support for each defined route,
 // only if there's no OPTIONS handler for that route yet
-func AutogenOptionsRoutes(routes []*rest.Route, gen HttpOptionsGenerator) []*rest.Route {
+func AutogenOptionsRoutes(router *gin.Engine, gen HttpOptionsGenerator) {
 
+	routes := router.Routes()
 	methodGroups := make(map[string][]string, len(routes))
 
 	for _, route := range routes {
-		methods, ok := methodGroups[route.PathExp]
+		if strings.HasPrefix(route.Path, "/api/internal") {
+			continue
+		}
+		methods, ok := methodGroups[route.Path]
 		if !ok {
 			methods = make([]string, 0)
 		}
 
-		methodGroups[route.PathExp] = append(methods, route.HttpMethod)
+		methodGroups[route.Path] = append(methods, route.Method)
 	}
 
-	options := make([]*rest.Route, 0, len(methodGroups))
 	for route, methods := range methodGroups {
 		// skip if there's a handler for OPTIONS already
 		if !supportsMethod(http.MethodOptions, methods) {
-			options = append(options,
-				rest.Options(route, gen(methods)))
+			router.OPTIONS(route, gen(methods))
 		}
 	}
 
-	return append(routes, options...)
+}
+
+func init() {
+	if mode := os.Getenv(gin.EnvGinMode); mode != "" {
+		gin.SetMode(mode)
+	} else {
+		gin.SetMode(gin.ReleaseMode)
+	}
+	gin.DisableConsoleColor()
+}
+
+func NewRouter(app inv.InventoryApp) http.Handler {
+	router := gin.New()
+	router.Use(accesslog.Middleware())
+	router.Use(requestid.Middleware())
+
+	mgmtHandler := NewManagementHandler(app)
+
+	publicAPIs := router.Group(".")
+
+	publicAPIs.Use(identity.Middleware())
+
+	mgmtAPIV1 := publicAPIs.Group(apiUrlManagementV1)
+	mgmtAPIV1Legacy := publicAPIs.Group(apiUrlLegacy)
+	mgmtAPIV2 := publicAPIs.Group(apiUrlManagementV2)
+	devicesAPIs := publicAPIs.Group(apiUrlDevicesV1)
+	devicesAPILegacy := publicAPIs.Group(apiUrlLegacy)
+
+	mgmtAPIV1.GET(uriDevices, mgmtHandler.GetDevicesHandler)
+	mgmtAPIV1.GET(uriDevice, mgmtHandler.GetDeviceHandler)
+	mgmtAPIV1.GET(uriDeviceGroups, mgmtHandler.GetDeviceGroupHandler)
+	mgmtAPIV1.GET(uriGroups, mgmtHandler.GetGroupsHandler)
+	mgmtAPIV1.GET(uriGroupsDevices, mgmtHandler.GetDevicesByGroupHandler)
+	rest.ApplyMiddlewareToRoutes(
+		contenttype.CheckJSON(),
+		mgmtAPIV1.DELETE(uriDevice, mgmtHandler.DeleteDeviceInventoryHandler),
+		mgmtAPIV1.DELETE(uriDeviceGroup, mgmtHandler.DeleteDeviceGroupHandler),
+		mgmtAPIV1.DELETE(uriGroupsName, mgmtHandler.DeleteGroupHandler),
+		mgmtAPIV1.DELETE(uriGroupsDevices, mgmtHandler.ClearDevicesGroupHandler),
+		mgmtAPIV1.PUT(uriDeviceGroups, mgmtHandler.AddDeviceToGroupHandler),
+		mgmtAPIV1.PATCH(uriGroupsDevices, mgmtHandler.AppendDevicesToGroup),
+		mgmtAPIV1.PUT(uriDeviceTags, mgmtHandler.UpdateDeviceTagsHandler),
+		mgmtAPIV1.PATCH(uriDeviceTags, mgmtHandler.UpdateDeviceTagsHandler),
+	)
```

Comment:
Sorry I didn't catch this in your previous PR earlier. I took another look at the gin reference and it turns out that the call to make routes (`GET`, `DELETE`, etc.) returns the group object itself. In turn, the call to `ApplyMiddlewareToRoutes` will apply the middleware as many times as there are arguments.

Instead you could do:
```suggestion
	mgmtAPIV1.DELETE(uriDevice, mgmtHandler.DeleteDeviceInventoryHandler),
	mgmtAPIV1.DELETE(uriDeviceGroup, mgmtHandler.DeleteDeviceGroupHandler),
	mgmtAPIV1.DELETE(uriGroupsName, mgmtHandler.DeleteGroupHandler),
	mgmtAPIV1.DELETE(uriGroupsDevices, mgmtHandler.ClearDevicesGroupHandler),

	mgmtAPIV1.Group(".").Use(contenttype.CheckJSON()).
		.PUT(uriDeviceGroups, mgmtHandler.AddDeviceToGroupHandler)
		.PATCH(uriGroupsDevices, mgmtHandler.AppendDevicesToGroup)
		.PUT(uriDeviceTags, mgmtHandler.UpdateDeviceTagsHandler)
		.PATCH(uriDeviceTags, mgmtHandler.UpdateDeviceTagsHandler)
```

Reply:
I created https://github.com/mendersoftware/mender-server/pull/656